### DISCUSSION TITLE
how to add an "alias" (function with spread args) to a plugin?

### DISCUSSION THREAD

**Discussion by JeffMelton** [nushell/nushell#13965] [Q&A] [answered]
I found the [prometheus](https://github.com/drbrain/nu_plugin_prometheus) plugin earlier and have been poking around a bit trying to define a function-alias `p8s` to save me a pile of keystrokes. Seemingly the most straightforward approach, placed in modules.nu as I've done with external commands:

```nu
export def --wrapped p8s [...args] {
    prometheus ...$args
}
```

returns 

```nu
Error: nu::parser::unexpected_spread_arg

  × This command does not have a ...rest parameter
    ╭─[/Users/<me>/Library/Application Support/nushell/lib/modules.nu:13:16]
 12 │ export def --wrapped p8s [...args] {
 13 │     prometheus ...$args
    ·                ────┬───
    ·                    ╰── unexpected spread argument
 14 │ }
    ╰────
```
when I open a new login shell. 

It seems that nushell treats plugins differently from external commands, so while the login shell will load without errors using `^prometheus` as if it's an external command, attempting to use the `p8s` function returns

```nu
Error: nu::shell::external_command

  × External command failed
    ╭─[/Users/j0m0aam/Library/Application Support/nushell/lib/modules.nu:13:6]
 12 │ export def --wrapped p8s [...args] {
 13 │     ^prometheus ...$args
    ·      ─────┬────
    ·           ╰── Command `prometheus` not found
 14 │ }
    ╰────
```

Am I missing something obvious here, or am I trying to do something that's not meant to be done? TIA

**Comment by fdncred** [comment, marked as answer]
You should probably file and issue on the author's repo. Plugins can't accept infinite arguments. They have to be written to support different types. I'm not sure if the plugin you mention has. 

/cc @drbrain

**Comment by ghost** [comment]
how about using `alias`? `alias k8s = prometheus` should do what your function aims to do

**Reply by JeffMelton** [reply]
@Jan9103 I'll look at it. In the past I've not been able to get plain aliases to take args, but I was admittedly very new to the shell then, so it's entirely possible that was a PEBKAC. 

Edit: Setting `alias p8s = prometheus` with my other aliases does what I remembered. Aliases don't take arguments, and I can't use it as a static command, as I have thousands of p8s instances.
```nu

  × Extra positional argument.
   ╭─[entry #2:1:30]
 1 │ "DCGM_FI_DEV_GPU_UTIL" | p8s query --url <redacted p8s endpoint>
   ·                              ──┬──
   ·                                ╰── extra positional argument
   ╰────
```

**Reply by ghost** [reply]
i checked a bit - `alias` does support arguments, but not having half command names (even if the command name contains a space - `from xml` is one command-name, not a command-name + argument).  
you would have to add `alias "p8s query" = prometheus query`, etc all individually..  
you could generate the aliases via

```nu
help commands | where $it.name =~ '^kubernetes ' | each {|i| $"alias \"k8s ($i.name | split row ' ' | get 1.. | str join ' ')\" = ($i.name)"} | str join "\n"
```

**Reply by JeffMelton** [reply]
Thanks for looking at that. I'll noodle with it. 😁

**Comment by drbrain** [comment]
Thanks for using my plugin!

I'd be happy to consider changes to arguments, or improved use of the [plugin config file](https://github.com/drbrain/nu_plugin_prometheus?tab=readme-ov-file#sources) beyond storing sources in the issues on that repository.

**Reply by JeffMelton** [reply]
Thanks for building it! I do a lot with Prometheus, so I was excited to find it. :) I'm not the person to offer suggestions for arguments or config, but I'll drop an issue over there about exporting `p8s` as well as `prometheus`.
